sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/TableIdentifier.scala

```diff
@@ -20,14 +20,20 @@ package org.apache.spark.sql.catalyst
 /**
  * Identifies a `table` in `database`.  If `database` is not defined, the current database is used.
  */
-private[sql] case class TableIdentifier(table: String, database: Option[String] = None) {
-  def withDatabase(database: String): TableIdentifier = this.copy(database = Some(database))
+private[sql] case class TableIdentifier(table: String, database: Option[String]) {
+  override def toString: String = {
+    if (table.contains('.') || database.exists(_.contains('.'))) {
```

Comment:
`.` is not the only character that need to be quoted, see http://stackoverflow.com/questions/4200351/what-characters-are-valid-in-an-sql-server-database-name